### PR TITLE
CXSPA-8150:adapt Product Carousel in SPA to use enhanced Products Search OCC API

### PR DESCRIPTION
**Pull request by davidabap**
Code change in `package.json` and `.env-cmdrc` just for local testing which link to Smartedit team internal system which have the latest Product Search API feature. These change will be revert before PR merge

### CODE REVIEW

File: .env-cmdrc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "dev": {
-    "CX_BASE_URL": "https://40.76.109.9:9002"
+    "CX_BASE_URL": "https://api.cspno58gmm-smartedit1-d5-public.model-t.myhybris.cloud"
```
Comment (pawelfras) [inline .env-cmdrc]: `major:` Let's not forget to revert this change

File: package.json [hunk 2]
```diff
@@ -91,7 +91,7 @@
     "lint:styles": "stylelint \"{projects,feature-libs}/**/*.scss\"",
     "prettier": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\"",
     "prettier:fix": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\" --write",
-    "start": "env-cmd --no-override -e dev,b2c,$SPA_ENV nx serve storefrontapp --configuration=development",
+    "start": "env-cmd --no-override -e dev,b2c,$SPA_ENV nx serve storefrontapp --configuration=development --ssl",
```
Comment (pawelfras) [inline package.json]: `major:` let's not forget to revert this change

File: package.json [hunk 3]
```diff
@@ -101,6 +101,7 @@
...
     "test:storefront:lib": "nx test storefrontlib --source-map --code-coverage",
+    "test:core:lib": "nx test core --source-map --code-coverage",
```
Comment (pawelfras) [inline package.json]: `major:` let's not forget to revert this change

File: projects/core/src/product/store/effects/product-search-by-code.effect.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  searchByCodes$ = createEffect(
+    () =>
+      ({ scheduler, debounce = 0 } = {}): Observable<
+        | ProductActions.ProductSearchLoadByCodeSuccess
+        | ProductActions.ProductSearchLoadByCodeFail
+      > =>
+        this.actions$.pipe(
+          ofType(ProductActions.PRODUCT_SEARCH_LOAD_BY_CODE),
+
+          // We split streams of actions by scope, so later we will be able
```
Comment (pawelfras) [inline projects/core/src/product/store/effects/product-search-by-code.effect.ts:44]: `nitpick:` I think we can remove comments describing the logic.

File: projects/core/src/product/store/effects/product-search-by-code.effect.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      this.actions$.pipe(
+        ofType(AuthActions.LOGOUT, AuthActions.LOGIN),
+        map(() => new ProductActions.ClearProductSearchByCodeState())
+      )
+    );
+
+  constructor(
+    private actions$: Actions,
+    private productSearchConnector: ProductSearchConnector
+  ) {}
```
Comment (pawelfras) [inline projects/core/src/product/store/effects/product-search-by-code.effect.ts]: `major:` Let's not combine DI provided by the constructor and `inject` function. To keep our coding guideline, let's use only the `inject` function to prevent breaking changes. See: https://wiki.one.int.sap/wiki/display/spar/Coding+Guidelines#CodingGuidelines-Useinject()functioninsteadofconstructorforDependencyInjection

```suggestion
  // at the top of the class
  private actions$ = inject(Actions);
  private productSearchConnector = inject(ProductSearchConnector);
```

File: projects/core/src/features-config/feature-toggles/config/feature-toggles.ts [hunk 6]
```diff
@@ -77,6 +77,11 @@ export interface FeatureTogglesInterface {
    */
   storeFrontLibCardParagraphTruncated?: boolean;
 
+  /**
+   * In `ProductCarouselComponent` it uses the batch API to load products.
+   */
```
Comment (pawelfras) [inline projects/core/src/features-config/feature-toggles/config/feature-toggles.ts]:
```suggestion
 /**
   * When enabled, the batch API is used `ProductCarouselComponent` to load products. It increases the component's performance.
   *
   * _NOTE_: When flag is enabled, custom OCC config for the `productSearch` endpoint has to be adjusted to have an object representation:
   * ```js
   * backend: {
   *    occ: {
   *      endpoints: {
   *         productSearch: {
   *           default: '...',
   *           carousel: '...',
   *           carouselMinimal: '...',
   *         },
   *       },
   *     },
   *   }
   * ```
   */
```

File: projects/core/src/product/facade/product-search-by-code.service.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import { Product } from '../../model';
+import { ProductActions } from '../store/actions/index';
+import { StateWithProduct } from '../store/product-state';
+import { ProductSelectors } from '../store/selectors/index';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ProductSearchByCodeService {
+  constructor(protected store: Store<StateWithProduct>) {}
```
Comment (pawelfras) [inline projects/core/src/product/facade/product-search-by-code.service.ts]: major: For DI, let's use the `inject` function to prevent breaking changes in the constructor signature if a new dependency is required. See: https://wiki.one.int.sap/wiki/display/spar/Coding+Guidelines#CodingGuidelines-Useinject()functioninsteadofconstructorforDependencyInjection

File: .env-cmdrc (hunk 1 above)
Comment (davidabap) [reply .env-cmdrc]: Done

File: package.json (hunk 2 above)
Comment (davidabap) [reply package.json]: Done

File: package.json (hunk 3 above)
Comment (davidabap) [reply package.json]: Done

File: projects/core/src/product/store/effects/product-search-by-code.effect.ts (hunk 5 above)
Comment (davidabap) [reply projects/core/src/product/store/effects/product-search-by-code.effect.ts]: Done

File: projects/core/src/features-config/feature-toggles/config/feature-toggles.ts (hunk 6 above)
Comment (davidabap) [reply projects/core/src/features-config/feature-toggles/config/feature-toggles.ts]: Done

File: projects/core/src/product/facade/product-search-by-code.service.ts (hunk 7 above)
Comment (davidabap) [reply projects/core/src/product/facade/product-search-by-code.service.ts]: Done

File: projects/core/src/product/store/effects/product-search-by-code.effect.ts (hunk 4 above)
Comment (davidabap) [reply projects/core/src/product/store/effects/product-search-by-code.effect.ts:44]: I think it's developer friendly for someone(Me) fresh to ngRx or rxjs

**Review by pawelfras — COMMENTED** [review]
LGTM code-wise. For QA, we need to update dev env with latest Commerce version.

**Review by pawelfras — APPROVED** [review]
LGTM. QA done using test environment.
As discussed offline, we will enable this feature for storefront app in a separate ticket (https://jira.tools.sap/browse/CXSPA-8162) when the dev environment contains latest Commerce version that supports changes from this PR
